src/main/java/org/elasticsearch/common/settings/ImmutableSettings.java

```diff
@@ -1100,6 +1102,17 @@ public boolean shouldIgnoreMissing(String placeholderName) {
         }
 
         /**
+         * Runs through all settings in this builder and prompts the user to provide a value for each <code>__prompt__</code>
+         * value that is found in the settings.
+         * @param terminal The terminal to user to for input/output
+         * @return The builder
+         */
+        public Builder replacePromptPlaceholders(Terminal terminal) {
```

Comment:
I don't think this class should have a method with this kind of sideeffects. I thinks we should put all this stuff into 
`PromptPlaceholderResolver` and have a method there with this signature:

```Java
public static Settings promtTerminal(Terminal terminal, Settings settings) {
  // this method never modifies a builder we take the settings and build a new one and return
  return newSettings;
}
```

this way we don't need all the static maps in that class either and we don't need to access inner structures. In `prepareSettings` we call this as the last thing before we return the settings as the tuple.

does thsi make sense?